lib/Filters/GitDiffReader.cpp

```diff
@@ -0,0 +1,79 @@
+#include "mull/Filters/GitDiffReader.h"
+
+#include "mull/Diagnostics/Diagnostics.h"
+#include "mull/Toolchain/Runner.h"
+
+#include <llvm/Support/FileSystem.h>
+
+#include <iostream>
+#include <regex>
+#include <sstream>
+
+using namespace mull;
+
+GitDiffReader::GitDiffReader(Diagnostics &diagnostics, const std::string &gitRepoPath)
+    : diagnostics(diagnostics), gitRepoPath(gitRepoPath) {}
+
+GitDiffInfo GitDiffReader::readGitDiff(const std::string &gitBranch) {
+  /// The implementation is borrowed from the git-clang-format Python tool.
+  /// https://opensource.apple.com/source/clang/clang-800.0.38/src/tools/clang/tools/clang-format/git-clang-format.auto.html
+  Runner runner(diagnostics);
+
+  std::vector<std::string> args = { "diff", "-U0", gitBranch };
+  std::vector<std::string> env = {};
+
+  ExecutionResult result = runner.runProgram("git", args, env, 1000, true);
+  if (result.exitStatus != 0) {
+    diagnostics.error(
+        std::string("GitDiffReader: cannot get git diff information. Received output: ") +
+        result.stderrOutput);
+    exit(1);
+  }
+
+  const std::string diffContent = result.stdoutOutput;
+  GitDiffInfo gitDiffInfo = parseDiffContent(diffContent);
+
+  return gitDiffInfo;
+}
+
+GitDiffInfo GitDiffReader::parseDiffContent(const std::string &diffContent) {
+  GitDiffInfo gitDiffInfo;
+  if (diffContent.empty()) {
+    return gitDiffInfo;
+  }
+
+  std::stringstream diffStream(diffContent);
+  std::string currentLine;
+
+  std::string currentFileName;
+
+  while (std::getline(diffStream, currentLine, '\n')) {
+    std::regex lineRegex("^\\+\\+\\+\\ [^/]+/(.*)");
+    std::regex rangeRegex("^@@ -[0-9,]+ \\+(\\d+)(,(\\d+))?");
+    std::smatch matches;
+
+    if (std::regex_search(currentLine, matches, lineRegex)) {
+      // TODO: What is the orthodox way of joining path components?
```

Comment:
Pretty much what TODO says.

Reply:
We have couple more places where "/" is hardcoded, so whatever :)